Issue #400: Simplify `add_pairwise_comparison()`

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #400.

`add_pairwise_comparison()` and `pairwise_comparison()` are quite complicated functions and some parts have become outdated/obsolete. This PR tries to simplify and clean them a bit. 

Changes in this PR: 
- Changes to `add_pairwise_comparison()`:
  - changes the default argument for `by` in `add_pairwise_comparison()` from `NULL` to `"model"`. `NULL` doesn't really make any sense, because if `by` is set to the forecast unit, you can't actually make any pairwise comparisons (as there is always only one forecast per forecast unit)
  - Removed the function `check_summary_params()` from `add_pairwise_comparison()` and moved all input checks to `pairwise_comparison()` directly where they are more imminently necessary. 
  - removes any stored attributes related to either unsummarised scores or any summary level stored (previously the summary level was stored in `scoringutils_by` as well as unsummarised scores). 
  - updates the names of the `score_names` stored to be in line with the recent change in `pairwise_comparison()` where we now have something like `wis_relative_skill` instead of just `relative_skill`
  - `add_pairwise_comparison()` no longer calls `summarise_scores()`, meaning that it returns unsummarised scores
- changes to `pairwise_comparison()`
  - added more input checks
  - simplified warnings and errors by getting rid of `paste0()`
  - removed old and unnecessary code that summarised across samples / quantiles. This is no longer necessary as the output of `score()` now naturally comes in that format (i.e. one score per forecast, rather than per sample/quantile)
- added extra tests for the input checks for `pairwise_comparison()`. Fixed some tests that were trying to compute pairwise comparisons based on "model" alone. 
- unrelated changes: none I think (🎉 )

Further thoughts and things to do (likely in a different PR): 
- Rename the `relative_skill_metric` argument (I'd do that together with all renaming of similar arguments) --> https://github.com/epiforecasts/scoringutils/issues/610
- Maybe we should just do away with the `"auto"`-thing and force users to spell out the name of the metric. Or reconsider whether the default should be called "auto"....
- `pairwise_comparison()` currently returns `NULL` and throws a warning if something goes wrong. Likely it should just error. --> https://github.com/epiforecasts/scoringutils/issues/637

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [ ] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [ ] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

seabbs commented:
> unrelated changes: none I think 

could this really be possible? I don't believe it!

seabbs commented:
> Rename the relative_skill_metric argument (I'd do that together with all renaming of similar arguments)

Good idea. 

> Maybe we should just do away with the "auto"-thing and force users to spell out the name of the metric. Or reconsider whether the default should be called "auto"....

Yeah agree. We could have the default be the data input with a grepl for relative skill on the column names? If there is more than one then make it error? (i.e do above in the input arg not in the code).

> pairwise_comparison() currently returns NULL and throws a warning if something goes wrong. Likely it should just error.

Definitely

seabbs commented on README.md:
```diff
@@ -171,17 +171,21 @@ example_quantile %>%
-| UMass-MechBayes       | Deaths      |    53 |            9.0 |            17.0 |         27 | -0.0220 |                 0.46 |                 0.88 |                      -0.025 |        78 |               0.75 |                       1.3 |
-| epiforecasts-EpiNow2  | Cases       | 21000 |        12000.0 |          3300.0 |       5700 | -0.0790 |                 0.47 |                 0.79 |                      -0.070 |     28000 |               0.95 |                       1.2 |
-| epiforecasts-EpiNow2  | Deaths      |    67 |           19.0 |            16.0 |         32 | -0.0051 |                 0.42 |                 0.91 |                      -0.045 |       100 |               0.98 |                       1.6 |
+| EuroCOVIDhub-baseline | Cases       | 28000 |        14000.0 |         10000.0 |       4100 |  0.0980 |                 0.33 |                 0.82 |                      -0.120 |     38000 |               1.60 |                      2.00 |
```
Looks like values are changing? Are they? Do we expect that? I think no?

nikosbosse replied:
Good catch! The reason they changed is because now that we removed the `unsummarised_scores` attribute the order of `add_pairwise_comarison()` and `summarise_scores()` matters. I fixed that.

seabbs reviewed (changes requested):
This is a really nice PR! I think everything looks good but just want to check that the changes in README output make sense.

Otherwise just needs a news update and promotion to issues the comments made about other work/features.

nikosbosse commented:
> We could have the default be the data input with a grepl for relative skill on the column names? If there is more than one then make it error? (i.e do above in the input arg not in the code).

You mean something like the following? 
```
pairwise_comparison <- function(scores,
                                by = "model",
                                metric = intersect(c("wis", "crps", "brier_score"), names(scores)),
                                baseline = NULL,
                                ...) {
```

seabbs reviewed (approved):
LGTM - just the auto setting to decide on

nikosbosse commented:
I'm fine with your default suggestion and made the following updates: 
- add_pairwise_comparison(): 
  - changed the default argument to `metric = intersect(c("wis", "crps", "brier_score"), names(scores))`
- `pairwise_comparison()`:
  - changed the default argument to `metric = intersect(c("wis", "crps", "brier_score"), names(scores))`
  - couldn't hold myself back from updating the following input check (finally at least an unrelated update!)
  ```
  (!is.null(baseline) && !(baseline %in% models)) {
    stop(
      "The baseline provided was not found among the models in the data."
    )
  }
  ```
  ```
  assert_subset(baseline, models)
  ```
  - and this one: 
  ```
  model_col <- check_columns_present(scores, "model")
    if (!is.logical(model_col)) {
      stop(
        "To compute relative skill, a column called 'model' ",
        "must be present in the input data."
      )
    }
  ```
  assert(check_columns_present(scores, "model"))
  ```
- updated tests to reflect that
- made rudimentary updates to the documentation. A complete update should be in a separate PR (see https://github.com/epiforecasts/scoringutils/issues/594)

nikosbosse commented:
@seabbs are you happy for me to merge this one as well?
